src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -867,56 +921,78 @@ private void WriteHeaders(HttpRequestMessage request)
             }
         }
 
-        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request)
+        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
             // Ensure we don't exceed the max concurrent streams setting.
-            await _concurrentStreams.RequestCreditAsync(1).ConfigureAwait(false);
+            await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
 
-            // Note, HEADERS and CONTINUATION frames must be together, so hold the writer lock across sending all of them.
-            // We also serialize usage of the header encoder and the header buffer this way.
-            // (If necessary, we could have a separate semaphore just for creating and encoding header blocks,
-            // and defer taking the actual _writerLock until we're ready to do the write below.)
-            await _writerLock.WaitAsync().ConfigureAwait(false);
+            // We serialize usage of the header encoder and the header buffer seperately from the
+            // write lock
+            await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-            Http2Stream http2Stream = AddStream(request);
-            int streamId = http2Stream.StreamId;
+            int streamId = 0;
+            int totalSize = 0;
+            Http2Stream http2Stream = null;
+            ReadOnlyMemory<byte> remaining = null;
+            ReadOnlyMemory<byte> current = null;
+            FrameFlags flags;
 
             try
```

Comment:
I need to handle cancellation differently when we're serializing the headers vs when we're actually writing to the wire. Splitting this up into two separate try-catch blocks feels a little bit awkward, but I wasn't sure how else to handle it. Thoughts?

Reply:
I don't think we can release the header lock before we do the write. Part of what it is protecting is that the stream IDs in headers blocks must always increase. If we release the header lock before the writer lock, we could do the writes out of order.

Reply:
Thanks. I'll fix that. I think it will actually simplify things a bit here.